File: PEUserMessageHandler/src/nl/adaptivity/process/userMessageHandler/server/XmlTask.java
package nl.adaptivity.process.userMessageHandler.server;

import java.security.Principal;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.xml.bind.JAXBException;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

import net.devrieze.util.security.SimplePrincipal;

import nl.adaptivity.messaging.EndpointDescriptorImpl;
import nl.adaptivity.messaging.MessagingException;
import nl.adaptivity.process.client.ServletProcessEngineClient;
import nl.adaptivity.process.exec.IProcessNodeInstance.TaskState;
import nl.adaptivity.process.util.Constants;

@XmlRootElement(name = "task")
@XmlAccessorType(XmlAccessType.NONE)
public class XmlTask implements UserTask<XmlTask> {

  private long aHandle = -1L;

  private long aRemoteHandle = -1L;

  private long aInstanceHandle = -1L;

  TaskState aState = TaskState.Sent;

  private String aSummary;

  private EndpointDescriptorImpl aEndPoint = null;

  private Principal aOwner;

  private List<XmlItem> aItems;

  public XmlTask() {
    // no special operations here
  }

  public XmlTask(final long pHandle) {
    aHandle = pHandle;
  }

  public XmlTask(UserTask<?> pTask) {
    this.aRemoteHandle = pTask.getRemoteHandle();
    this.aInstanceHandle = pTask.getInstanceHandle();
    this.aState = pTask.getState();
    this.aSummary = pTask.getSummary();
    this.aEndPoint = null;
    this.aOwner = pTask.getOwner();
    this.aItems = new ArrayList<>(XmlItem.get(pTask.getItems()));
  }

  @XmlAttribute
  @Override
  public TaskState getState() {
    return aState;
  }

  void setState(TaskState pState) {
    aState = pState;
  }

  @Override
  public void setState(final TaskState pNewState, final Principal pUser) {
    try {
      TaskState newState;
      if (pNewState == TaskState.Complete) {
        newState = finishRemoteTask(pUser).get();
        //          newState = TaskState.Complete; // Use server state instead.
      } else if (pNewState == TaskState.Acknowledged) {
        newState = pNewState; // Just shortcircuit. This is just record keeping
      } else {
        newState = updateRemoteTaskState(pNewState, pUser).get();
      }
      aState = newState;
    } catch (final JAXBException e) {
      Logger.getLogger(getClass().getCanonicalName()).throwing("XmlTask", "setState", e);
    } catch (final MessagingException e) {
      Logger.getLogger(getClass().getCanonicalName()).throwing("XmlTask", "setState", e);
    } catch (final InterruptedException e) {
      Logger.getAnonymousLogger().log(Level.INFO, "Messaging interrupted", e);
    } catch (final ExecutionException e) {
      Logger.getAnonymousLogger().log(Level.WARNING, "Error updating task", e);
    }
  }

  private Future<TaskState> updateRemoteTaskState(final TaskState pState, final Principal pUser) throws JAXBException, MessagingException {
    return ServletProcessEngineClient.updateTaskState(aRemoteHandle, pState, pUser, null);
  }

  private Future<TaskState> finishRemoteTask(final Principal pUser) throws JAXBException, MessagingException {
    return ServletProcessEngineClient.finishTask(aRemoteHandle, createResult(), pUser, null); // Ignore completion???
  }

  private Node createResult() {
    DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
    dbf.setNamespaceAware(true);;
    Document document;
    try {
      document = dbf.newDocumentBuilder().newDocument();
    } catch (ParserConfigurationException e) {
      throw new RuntimeException(e);
    }

    Element outer = document.createElementNS(Constants.USER_MESSAGE_HANDLER_NS, "result");
    for(XmlItem item: getItems()) {
      if (! "label".equals(item.getType())) {
        Element inner = document.createElementNS(Constants.USER_MESSAGE_HANDLER_NS, "value");
        inner.setAttribute("name", item.getName());
        if (item.getValue()!=null) {
          inner.setTextContent(item.getValue());
        }
        outer.appendChild(inner);
      }
    }

    return outer;
  }

  @XmlAttribute(name = "handle")
  @Override
  public void setHandle(final long pHandle) {
    aHandle = pHandle;
  }

  @Override
  public long getHandle() {
    return aHandle;
  }

  @XmlAttribute(name = "remotehandle")
  public void setRemoteHandle(final long pHandle) {
    aRemoteHandle = pHandle;
  }

  @Override
  public long getRemoteHandle() {
    return aRemoteHandle;
  }

  @XmlAttribute(name = "instancehandle")
  public void setInstanceHandle(final long pHandle) {
    aInstanceHandle = pHandle;
  }

  @Override
  public long getInstanceHandle() {
    return aInstanceHandle;
  }

  @Override
  @XmlAttribute(name = "summary")
  public String getSummary() {
    return aSummary;
  }

  public void setSummary(final String pSummary) {
    aSummary = pSummary;
  }

  /** Set the endpoint that is used for updating the task state */
  @Override
  public void setEndpoint(final EndpointDescriptorImpl pEndPoint) {
    aEndPoint = pEndPoint;
  }

  @Override
  public Principal getOwner() {
    return aOwner;
  }

  public void setOwner(final Principal pOwner) {
    aOwner = pOwner;
  }

  @XmlAttribute(name = "owner")
  String getOwnerString() {
    return aOwner==null ? null : aOwner.getName();
  }

  void setOwnerString(final String pOwner) {
    aOwner = pOwner==null ? null : new SimplePrincipal(pOwner);
  }

  @XmlElement(name ="item", namespace=Constants.USER_MESSAGE_HANDLER_NS)
  @Override
  public List<XmlItem> getItems() {
    if (aItems==null) { aItems = new ArrayList<>(); }
    return aItems;
  }

  @Override
  public void setItems(List<? extends TaskItem> pItems) {
    aItems = new ArrayList<>(pItems.size());
    for(TaskItem item: pItems) {
      aItems.add((XmlItem) item);
    }
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    result = prime * result + ((aEndPoint == null) ? 0 : aEndPoint.hashCode());
    result = prime * result + (int) (aHandle ^ (aHandle >>> 32));
    result = prime * result + ((aItems == null||aItems.isEmpty()) ? 0 : aItems.hashCode());
    result = prime * result + ((aOwner == null) ? 0 : aOwner.hashCode());
    result = prime * result + (int) (aRemoteHandle ^ (aRemoteHandle >>> 32));
    result = prime * result + ((aState == null) ? 0 : aState.hashCode());
    result = prime * result + ((aSummary == null) ? 0 : aSummary.hashCode());
    return result;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    XmlTask other = (XmlTask) obj;
    if (aEndPoint == null) {
      if (other.aEndPoint != null)
        return false;
    } else if (!aEndPoint.equals(other.aEndPoint))
      return false;
    if (aHandle != other.aHandle)
      return false;
    if (aItems == null || aItems.isEmpty()) {
      if (other.aItems != null && ! aItems.isEmpty())
        return false;
    } else if (!aItems.equals(other.aItems))
      return false;
    if (aOwner == null) {
      if (other.aOwner != null)
        return false;
    } else if (!aOwner.equals(other.aOwner))
      return false;
    if (aRemoteHandle != other.aRemoteHandle)
      return false;
    if (aState != other.aState)
      return false;
    if (aSummary == null) {
      if (other.aSummary != null)
        return false;
    } else if (!aSummary.equals(other.aSummary))
      return false;
    return true;
  }

  public static XmlTask get(UserTask<?> pTask) {
    if (pTask instanceof XmlTask) { return (XmlTask) pTask; }
    return new XmlTask(pTask);
  }

  public XmlItem getItem(String pName) {
    for(XmlItem item: getItems()) {
      if (pName.equals(item.getName())) {
        return item;
      }
    }
    return null;
  }


}

File: PEUserMessageHandler/src/main/kotlin/nl/adaptivity/process/userMessageHandler/server/XmlTask.kt
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.userMessageHandler.server

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import net.devrieze.util.Handle
import net.devrieze.util.HandleSerializer
import net.devrieze.util.collection.replaceBy
import net.devrieze.util.security.SimplePrincipal
import nl.adaptivity.messaging.EndpointDescriptorImpl
import nl.adaptivity.messaging.MessagingException
import nl.adaptivity.process.client.ServletProcessEngineClient
import nl.adaptivity.process.engine.processModel.NodeInstanceState
import nl.adaptivity.process.processModel.AccessRestriction
import nl.adaptivity.process.util.Constants
import nl.adaptivity.util.multiplatform.PrincipalCompat
import nl.adaptivity.xmlutil.*
import nl.adaptivity.xmlutil.serialization.XML
import nl.adaptivity.xmlutil.serialization.XmlElement
import nl.adaptivity.xmlutil.serialization.XmlSerialName
import org.w3c.dom.Document
import org.w3c.dom.DocumentFragment
import java.security.Principal
import java.util.concurrent.ExecutionException
import java.util.concurrent.Future
import java.util.logging.Level
import java.util.logging.Logger
import javax.xml.XMLConstants
import javax.xml.bind.JAXBException
import javax.xml.namespace.QName
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

@Serializable
@XmlSerialName("task", "http://adaptivity.nl/userMessageHandler", "umh")
class XmlTask : UserTask<XmlTask> {

    @Serializable(HandleSerializer::class)
    override var handle = Handle.invalid<XmlTask>()
        private set

    override val handleValue: Long
        get() = handle.handleValue

    override fun setHandleValue(handleValue: Long) {
        if (handle.handleValue != handleValue) this.handle =
            if (handleValue < 0) Handle.invalid() else Handle(handleValue)
    }

    @Serializable(HandleSerializer::class)
    @SerialName("remotehandle")
    override var remoteHandle: Handle<Unit> = Handle.invalid()

    @Serializable(HandleSerializer::class)
    @SerialName("instancehandle")
    override var instanceHandle: Handle<Unit> = Handle.invalid()

    // TODO make this not-nullable
    @XmlElement(false)
    override var state: NodeInstanceState? = NodeInstanceState.Sent

    override var summary: String? = null

    private var endPoint: EndpointDescriptorImpl? = null

    @Serializable(AccessRestrictionSerializer::class)
    @SerialName("restrictTo")
    @XmlElement(false)
    override var accessRestriction: AccessRestriction? = null

    @Serializable(PrincipalNameSerializer::class)
    override var owner: PrincipalCompat? = null

    @SerialName("item")
    private val _items = mutableListOf<XmlItem>()

    /*, namespace=Constants.USER_MESSAGE_HANDLER_NS*/
    override var items: List<XmlItem>
        get() = _items
        set(value) {
            _items.replaceBy(value)
        }

    constructor()

    constructor(handle: Long) {
        setHandleValue(handle)
    }

    constructor(task: UserTask<*>) : this(task.handleValue) {
        @Suppress("UNCHECKED_CAST")
        this.remoteHandle = task.remoteHandle as Handle<Unit>
        @Suppress("UNCHECKED_CAST")
        this.instanceHandle = task.instanceHandle as Handle<Unit>
        this.state = task.state
        this.summary = task.summary
        this.endPoint = null
        this.owner = task.owner
        this.items = XmlItem.get(task.items)
    }

    override fun setState(newState: NodeInstanceState, user: Principal) { // TODO handle transactions
        try {
            state = when (newState) {
                NodeInstanceState.Complete     -> finishRemoteTask(user).get()
                NodeInstanceState.Cancelled    -> cancelRemoteTask(user).get()
                NodeInstanceState.Acknowledged -> newState // Just shortcircuit. This is just record keeping
                else                           -> updateRemoteTaskState(newState, user).get()
            }
        } catch (e: XmlException) {
            Logger.getLogger(javaClass.canonicalName).throwing("XmlTask", "setState", e)
        } catch (e: JAXBException) {
            Logger.getLogger(javaClass.canonicalName).throwing("XmlTask", "setState", e)
        } catch (e: MessagingException) {
            Logger.getLogger(javaClass.canonicalName).throwing("XmlTask", "setState", e)
        } catch (e: InterruptedException) {
            Logger.getAnonymousLogger().log(Level.INFO, "Messaging interrupted", e)
        } catch (e: ExecutionException) {
            Logger.getAnonymousLogger().log(Level.WARNING, "Error updating task", e)
        }

    }

    @Throws(JAXBException::class, MessagingException::class, XmlException::class)
    private fun updateRemoteTaskState(state: NodeInstanceState, user: Principal): Future<NodeInstanceState> {
        return ServletProcessEngineClient.updateTaskState(remoteHandle.handleValue, state, user, null)
    }

    @Throws(JAXBException::class, MessagingException::class, XmlException::class)
    private fun finishRemoteTask(user: Principal): Future<NodeInstanceState> {
        return ServletProcessEngineClient.finishTask(
            remoteHandle.handleValue,
            createResult(),
            user,
            null
        ) // Ignore completion???
    }

    @Throws(JAXBException::class, MessagingException::class, XmlException::class)
    private fun cancelRemoteTask(user: Principal): Future<NodeInstanceState> {
        // For now don't do anything special
        return updateRemoteTaskState(NodeInstanceState.Cancelled, user)
    }

    private fun createResult(): DocumentFragment {
        val dbf = DocumentBuilderFactory.newInstance()
        dbf.isNamespaceAware = true
        val document: Document
        try {
            document = dbf.newDocumentBuilder().newDocument()
        } catch (e: ParserConfigurationException) {
            throw RuntimeException(e)
        }

        val fragment = document.createDocumentFragment()

        val outer = document.createElementNS(Constants.USER_MESSAGE_HANDLER_NS, "result")
        outer.setAttributeNS(XMLConstants.XMLNS_ATTRIBUTE_NS_URI, "xmlns", Constants.USER_MESSAGE_HANDLER_NS)
        for (item in items) {
            if ("label" != item.type) {
                val inner = document.createElementNS(Constants.USER_MESSAGE_HANDLER_NS, "value")
                inner.setAttribute("name", item.name)
                if (item.value != null) {
                    inner.textContent = item.value
                }
                outer.appendChild(inner)
            }
        }

        fragment.appendChild(outer)
        return fragment
    }

    /** Set the endpoint that is used for updating the task state  */
    override fun setEndpoint(endPoint: EndpointDescriptorImpl) {
        this.endPoint = endPoint
    }

    fun getItem(name: String) = items.firstOrNull { name == it.name }
    operator fun get(name: String) = getItem(name)

    override fun hashCode(): Int {
        var result = handle.hashCode()
        result = 31 * result + remoteHandle.handleValue.hashCode()
        result = 31 * result + instanceHandle.handleValue.hashCode()
        result = 31 * result + state.hashCode()
        result = 31 * result + summary.hashCode()
        result = 31 * result + endPoint.hashCode()
        result = 31 * result + accessRestriction.hashCode()
        result = 31 * result + owner.hashCode()
        result = 31 * result + _items.hashCode()
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as XmlTask

        if (handle != other.handle) return false
        if (remoteHandle != other.remoteHandle) return false
        if (instanceHandle != other.instanceHandle) return false
        if (state != other.state) return false
        if (summary != other.summary) return false
        if (endPoint != other.endPoint) return false
        if (accessRestriction != other.accessRestriction) return false
        if (owner != other.owner) return false
        if (_items != other._items) return false

        return true
    }

    companion object {

        val ELEMENTLOCALNAME = "task"
        val ELEMENTNAME = QName(Constants.USER_MESSAGE_HANDLER_NS, ELEMENTLOCALNAME, "umh")

        @Throws(XmlException::class)
        fun deserialize(xmlReader: XmlReader): XmlTask {
            return XML.decodeFromReader<XmlTask>(xmlReader)
        }

        @JvmStatic
        fun get(task: UserTask<*>): XmlTask {
            if (task is XmlTask) {
                return task
            }
            return XmlTask(task)
        }
    }

    internal class PrincipalNameSerializer: KSerializer<Principal> {
        override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Principal", PrimitiveKind.STRING)

        override fun serialize(encoder: Encoder, value: Principal) {
            encoder.encodeString(value.name)
        }

        override fun deserialize(decoder: Decoder): Principal {
            return SimplePrincipal(decoder.decodeString())
        }
    }

    internal class AccessRestrictionSerializer: KSerializer<AccessRestriction> {
        override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("AccessRestriction", PrimitiveKind.STRING)

        override fun serialize(encoder: Encoder, value: AccessRestriction) {
            encoder.encodeString(value.serializeToString())
        }

        override fun deserialize(decoder: Decoder): AccessRestriction {
            val str = decoder.decodeString().trim()
            return when(str[0]) {
                '@' -> when(str[1]) {
                    '[' -> {
                        check(str.last()==']')
                        str.substring(2, str.length-1).splitToSequence(",").map { it.trim() }.let { roles ->
                            RoleRestriction(roles.toSet())
                        }
                    }

                    else -> RoleRestriction(str.substring(1).trimStart())
                }
                '[' -> {
                    check(str.last()==']')
                    str.substring(1, str.length-1).splitToSequence(",").map { it.trim() }.let { users ->
                        UserRestriction(users.toSet())
                    }
                }
                else -> UserRestriction(str)
            }
        }
    }

}

